File: 8kyu/(8 kyu) MakeUpperCase/(8 kyu) MakeUpperCase.js
// 1 Plain solution
function makeUpperCase(str) {
    let res = '';

    for (let i = 0; i < str.length; i++) {
        const cc = str.charCodeAt(i);
        const c = String.fromCharCode(cc - 32);
        // Convert symbols for a..z only
        res += cc <= 122 && cc >= 97 ? c : str[i];
    }

    return res;
}
// 2 Straightforward solution
function makeUpperCase(str) {
    return str
        .replace(
            /[a-z]/g,
            s => String.fromCharCode(
                s.charCodeAt(0) - 32
            )
        )
}
// 3 Optimized solution
function makeUpperCase(str) {
    return str.toUpperCase();
}
// 4 Coding golf
makeUpperCase = s => s.toUpperCase()

File: 8kyu/(8 kyu) MakeUpperCase/(8 kyu) MakeUpperCase.dart
// 1 Plain solution
String makeUpperCase(String str) {
  var res = "";

  for (int i = 0; i < str.length; i++) {
    final cc = str.codeUnitAt(i);
    final c = String.fromCharCode(cc - 32);
    // Convert symbols for a..z only
    res += cc <= 122 && cc >= 97 ? c : str[i];
  }

  return res;
}

// 2 Optimized solution
String makeUpperCase(String str) {
  return str.replaceAllMapped(
    RegExp(r'[a-z]'),
    (match) => match.group(0)!.toUpperCase(),
  );
}

// 4 Clever solution
String makeUpperCase(String str) {
  return str.toUpperCase();
}

// 5 Coding golf
makeUpperCase(s) => s.toUpperCase();
